### PR TITLE
nrf52/i2c: Use mutex and IRQ for blocking

### PR DESCRIPTION
**Pull request by bergzand**
### Contribution description

This commit enhances the I2C code of the nRF52 family to block on a
mutex while the I2C transfer is busy. The mutex is unlocked in the ISR
when it is trigger by either a stop condition or an error condition.

While the current code already allows the MCU to sleep during the transfer, this enhancement allows the RIOT scheduler to schedule other (lower priority) threads during the transfer.

### Testing procedure

I2C operations should still work, any connected peripheral such as an BME280 must still function as before. 

@MrKevinWeiss Do you have the time and hardware to give this a shot with RIOT-HIL?

### Issues/PRs references

None

### CODE REVIEW

File: cpu/nrf52/periph/i2c.c [hunk 1]
```diff
@@ -94,6 +105,8 @@ void i2c_init(i2c_t dev)
     bus(dev)->PSEL.SCL = i2c_config[dev].scl;
     bus(dev)->PSEL.SDA = i2c_config[dev].sda;
     /* configure dev clock speed */
```
Comment (benpicco) [inline cpu/nrf52/periph/i2c.c:107]: That comment is misplaced now ;)

Comment (bergzand) [reply cpu/nrf52/periph/i2c.c:107]: Fixed, Thanks!

**Comment by haukepetersen** [comment]
Did you do some kind of runtime measurements for small i2c transfers? The reason why I back then implemented the wait loop like this was that the context switching overhead did not justify the benefits gained... 

In many use cases we see rather short transfers (e.g. read status reg from a device). For those we might end up in significant overhead: 
- mutx_lock() puts the calling thread to sleep
- context store
- scheduler selects internmediate thread to run
- context restore
- other thread runs (potentially very shortly)
- I2C ISR, mutex unlock
- context store
- scheduler run, selecting our initial thread again
- context restore
- end of 'wait loop'
--> So in bad cases, we get two context switches while the intermediate task did potentially very little to no runtime for short I2C payloads. So would be good to make sure its worth it :-)

**Comment by benpicco** [comment]
If there is a significant overhead for small transfers, we can just so what was proposed for DMA and fall back to polling if the transfer size is below a certain threshold.

The lpc2387 i2c driver uses the same pattern as proposed in this PR btw.

**Comment by haukepetersen** [comment]
Yes, absolutely. Then its just a manageable code size overhead :-)

**Comment by bergzand** [comment]
> Did you do some kind of runtime measurements for small i2c transfers?

I did now. My original assumption was that a register read with I2C is relative slow due to the need to transfer the address, then write the register address, restart the transmission with the address and finally read the byte.
I measured the code difference with a logic analyzer and the bmx280 test application on the nrf52dk board, setting a gpio before the blocking bits and clearing it after the blocking bits. This results in the following measurements:

| Type | Master | #14165  |
| ------------- |:-------------:| -----:|
| Read single register | 449.8us | 462.9us |
| Write single register | 336.5us | 352.9us |
| Read long register (24 bytes) | 2.516ms | 2.532ms |

The read long register is 24 bytes of calibration data in the bme280. Note that the default I2C speed on the nrf52dk board uses 100kHz for the I2C clock. I can redo the test on 400kHz if wanted.

**Comment by bergzand** [comment]
Furthermore, on the nrf52dk, the `tests/bench_mutex_pingpong` shows `{ "result" : 122841 }`, or 61420.5 mutex lock-unlocks per second, or 16.28 us per mutex lock-unlock

**Comment by haukepetersen** [comment]
Awesome, exactly the numbers I was looking for. For pure runtime overhead, we are looking at a ~4,5% penalty for short transfers, that is something we could live with I'd say.

Looking at the energy consumption aspect it seems to me we loose a light bit of efficiency in the same order of magnitude: the code in master should put the CPU into light sleep for the complete wait duration, while with #14165  the idle thread is only active for the transfer duration minus the time taken for the task switching, right?

So I agree that this PR is a good idea and the benefits outweight those (small) penalties.

**Comment by benpicco** [comment]
Feel free to squash btw :wink:

**Comment by bergzand** [comment]
> Feel free to squash btw wink

Squashed

**Comment by bergzand** [comment]
> Looking at the energy consumption aspect it seems to me we loose a light bit of efficiency in the same order of magnitude: the code in master should put the CPU into light sleep for the complete wait duration, while with #14165 the idle thread is only active for the transfer duration minus the time taken for the task switching, right?

Exactly, with an otherwise idle cpu, it should still light sleep for the transfer duration, with a small bit of extra overhead caused by the mutex in this PR.

I'm starting to think if switching between the `nrf52_sleep()`-option and the mutex based on a compile-time setting makes sense. Let the user decide if he/she really needs the amount of power saved by instantly sleeping.

**Comment by benpicco** [comment]
> Let the user decide if he/she really needs the amount of power saved by instantly sleeping.

Having two code paths just doubles the potential for error.
And if we can't decide on that, putting the burden on the user will not solve the problem.

File: cpu/nrf52/periph/i2c.c [hunk 2]
```diff
@@ -45,17 +45,25 @@
...
 {
     return i2c_config[dev].dev;
 }
 
 static int finish(i2c_t dev)
 {
+    /* Unmask interrupts */
+    bus(dev)->INTENSET = TWIM_INTEN_STOPPED_Msk | TWIM_INTEN_ERROR_Msk;
+    mutex_lock(&busy[dev]);
     DEBUG("[i2c] waiting for STOPPED or ERROR event\n");
```
Comment (benpicco) [inline cpu/nrf52/periph/i2c.c:66]: Should be moved before the lock - just squash again

Comment (bergzand) [reply cpu/nrf52/periph/i2c.c]: Fixed!

File: cpu/nrf52/periph/i2c.c [hunk 3]
```diff
@@ -53,9 +61,9 @@ static inline NRF_TWIM_Type *bus(i2c_t dev)
 static int finish(i2c_t dev)
 {
     DEBUG("[i2c] waiting for STOPPED or ERROR event\n");
```
Comment (benpicco) [inline cpu/nrf52/periph/i2c.c:63]: If I were to nitpick I'd say this should be before the `mutex_lock()` :wink:

**Comment by aabadie** [comment]
Tested with a couple of I2C sensors on adafruit-clue and still works.

File: cpu/nrf52/periph/i2c.c (hunk 3 above)
Comment (bergzand) [reply cpu/nrf52/periph/i2c.c:63]: I was thinking the same, but I didn't want to put a relative slow `DEBUG` statement between unmasking the IRQ and the lock.

**Comment by MrKevinWeiss** [comment]
For what it is work there is no change in i2c [test results](https://ci.riot-os.org/hil/blue/organizations/jenkins/RIOT-HIL/detail/master/2/pipeline/877/) from this pr and nightlies.

**Comment by bergzand** [comment]
> For what it is worth there is no change in i2c [test results](https://ci.riot-os.org/hil/blue/organizations/jenkins/RIOT-HIL/detail/master/2/pipeline/877/) from this pr and nightlies.

Thanks!

**Comment by bergzand** [comment]
Thanks for all the reviews and tests!

**Pull request merged**